sycl/doc/design/MappingHostAddressesToDeviceEntities.md

```diff
@@ -0,0 +1,240 @@
+When DPC++ compiler is used as both host and device compiler, then the attribute
+should be respected by both host and device compiler passes and LLVM IR
+attributes should appear in LLVM IR for both host and device code. When DPC++
+compiler is only used as a device compiler, then we don't expect the attribute
+to be handled on host, apparently.
+
+Another thing we need from DPC++ FE compiler is to define a special macro, which
+will allow to distinguish it from other compilers. That is needed to apply the
+aforementioned attribute conditionally to avoid spamming users with warnings
+about unknown attributes.
+
+The suggested macro name is `__INTEL_SYCL_HOST_COMPILER__`. It should be defined
+when the compiler is invoked in SYCL host mode (`-fsycl-is-host` `-cc1` flag).
+
+## Common headers part
+
+Header files should be modified by adding the new attributes to types
+declarations, objects of which we will need in our mapping.Again,
+`device_global` and `specialization_id` are examples here:
+
+```
+template <typename T>
+class
+#if defined(__SYCL_DEVICE_ONLY__) || defined(__INTEL_SYCL_HOST_COMPILER__)
+  [[__sycl_detail__::uniquely_identifiable_object("specialization_id")]]
+#endif
+specialization_id {
+// ...
+};
+```
+
+## Common runtime part
+
+The runtime should implement the following function, which will be called from
+a code generated by the compiler (see the next section):
+
+```
+void __register_uniquely_identifiable_object(
+  void *Address, const char* UniqueID, const char *Kind);
+```
+
+The function accepts the following arguments:
+- `Address` is an address of a variable, which exists in an application on host
+- `UniqueID` is a unique symbolic ID, which corresponds to that variable
+- `Kind` is a string which corresponds to `kind` argument passed to
+  `[[__sycl_detail__::uniquely_identifiable_object(kind)]]` attribute attached
+  to the type of the variable identified by `Address`. It can be used to
+  distinguish different entities like `specialization_id` and `device_global`:
+  for example they could be stored in different maps to speed up certain
+  operations with them.
+
+The compiler guarantees that the function will be called zero or more times
+(depending on the amount of uniquely identifiable objects found in a program)
+_before_ application's `main()` function, i.e. in a global constructor.
+
+That poses some restrictions on those uniquely identifiable object, i.e. that
+they can't be used from another global object due to risk of accessing a
+non-initialized object, but that is an UB anyway because the order of global
+objects initialization is not defined in C++ when those objects are defined in
+different translation unit.
+
+## Compiler driver part
+
+The compiler driver is the component which is responsible for selecting the
+approach we are taking and the decision is made based on whether or not
+3rd-party host compiler is in use.
+
+If `-fsycl-host-compiler` option is present, the compiler driver chooses the
```

Comment:
What will happen if `-fsycl-host-compiler=clang++` is passed?

Reply:
I honestly don't know what is the algorithm of determining an absolute path to a 3rd-party host compiler specified through that argument, but I suggest that we always assume that any compiler passed through the flag is a 3rd-party one: I simply don't think that it worth to implement compiler detection here, because even `clang++` can point to the upstream clang, for example and it doesn't have customizations we need.